reset.c

```diff
@@ -0,0 +1,858 @@
+            resetconf.keepRGs = 0;
+        break;
+        case 'p':                           //--no-PG [reject pg file]
+        {
+            if (resetconf.keepPGs <= 0) {
+                usage(stderr);              //already given!
+                return EXIT_FAILURE;
+            }
+            if (optarg) {                   //filename given, reject selected PGs only
+                pgrejname = optarg;
+                resetconf.keepPGs = 0;
+            }
+            else {
+                resetconf.keepPGs = -1;     //reject all PGs -- reset execution will anyway be added
+            }
+        }
+        break;
+        case 'i':                           //input file name
+        {
+            if (inname) {                   //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+            }
+            inname = optarg;
+        }
+        break;
+        case 'o':                           //output file name
+        {
+            if (outname) {                  //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+
+            }
+            outname = optarg;
+        }
+        break;
+        case 'k':                           //keep aux tags file name
+        {
+            if (auxkeepname) {              //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+
+            }
+            auxkeepname = optarg;
+        }
+        break;
+        case 'r':                           //rejected aux tags file name
+        {
+            if (auxrejname) {               //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+            }
+            auxrejname = optarg;
+        }
+        break;
+        case '@':                           //thread count
+        {
+            if (ga.nthreads != 0) {
+                usage(stderr);
+                return EXIT_FAILURE;
+            }
+            ga.nthreads = atoi(optarg);
+        }
+        break;
+        // handle standard samtool options like thread count, verbosity...
+        default:
+            if (parse_sam_global_opt(c, optarg, lopts, &ga) == 0)
+                break;
+            // else fall-through
+        // couldn't parse or unknown options, show usage!
+        case '?':                           //unknown options found!
+        {
+            usage(stderr);
+            return EXIT_FAILURE;
+            break;
+        }
+
+        }
+    }
+
+    //check and fail if unnecessary parameters are given
+    if ((argc - optind) > 0 || !inname || !outname)
+    {
+        usage(stderr);
+        return EXIT_FAILURE;
+    }
+
+    //get the aux tags to be retained
+    if (getauxtags(&resetconf, auxkeepname, auxrejname))
+        return EXIT_FAILURE;
+
+    //get the pg names to be discarded
+    if (getrejectedpgnames(&resetconf, pgrejname))
+        return EXIT_FAILURE;
```

Comment:
These can leak memory and so need a call to cleanup too.  To ensure cleanup doesn't try to do access memory it hasn't yet allocated, extra guards need adding there.

```
diff --git a/reset.c b/reset.c
index 6e6a8ce..85c2274 100644
--- a/reset.c
+++ b/reset.c
@@ -668,11 +668,13 @@ void cleanup(conf_data *config)
     khiter_t iter = 0;
     const char *tmp = NULL;
     kh_destroy(aux_set, config->auxtagsin);
-    for (iter = kh_begin(config->pgout); iter != kh_end(config->pgout); ++iter) {
-        tmp = (kh_exist(config->pgout,iter)) ? kh_key(config->pgout, iter) : NULL;
-        if (tmp)
-            free((void*)tmp);
-        tmp = NULL;
+    if (config->pgout) {
+        for (iter = kh_begin(config->pgout); iter != kh_end(config->pgout); ++iter) {
+            tmp = (kh_exist(config->pgout,iter)) ? kh_key(config->pgout, iter) : NULL;
+            if (tmp)
+                free((void*)tmp);
+            tmp = NULL;
+        }
     }
     kh_destroy(pg_set, config->pgout);
 }
@@ -802,12 +804,16 @@ int main_reset(int argc, char *argv[])
     }
 
     //get the aux tags to be retained
-    if (getauxtags(&resetconf, auxkeepname, auxrejname))
+    if (getauxtags(&resetconf, auxkeepname, auxrejname)) {
+        cleanup(&resetconf);
         return EXIT_FAILURE;
+    }
 
     //get the pg names to be discarded
-    if (getrejectedpgnames(&resetconf, pgrejname))
+    if (getrejectedpgnames(&resetconf, pgrejname)) {
+        cleanup(&resetconf);
         return EXIT_FAILURE;
+    }
 
     //set output file format based on name
     sam_open_mode(outmode + 1, outname, NULL);
```